core/ledger/kvledger/snapshot_test.go

```diff
@@ -339,6 +320,132 @@ func TestSnapshotCouchDBIndexCreation(t *testing.T) {
 		require.Len(t, actualResults, 1)
 		_, err = qe.ExecuteQuery("ns", `{"selector":{"owner":"tom"}, "sort": [{"color": "desc"}]}`)
 		require.Contains(t, err.Error(), "No index exists for this sort")
+	}
+
+	t.Run("create_indexes_on_couchdb_for_new_lifecycle", func(t *testing.T) {
+		snapshotDir, couchDBConfig, provider := setup()
+		defer func() {
+			require.NoError(t, statecouchdb.DropApplicationDBs(couchDBConfig))
+		}()
+
+		// mimic new lifecycle chaincode "ns" installed and defiend and the package contains an index definition "sort index"
+		ccLifecycleEventProvider := provider.initializer.ChaincodeLifecycleEventProvider.(*mock.ChaincodeLifecycleEventProvider)
+		ccLifecycleEventProvider.RegisterListenerStub =
+			func(
+				channelID string,
+				listener ledger.ChaincodeLifecycleEventListener,
+				callback bool,
+			) error {
+				if callback {
+					err := listener.HandleChaincodeDeploy(
+						&ledger.ChaincodeDefinition{
+							Name: "ns",
+						},
+						dbArtifactsBytes,
+					)
+					require.NoError(t, err)
+				}
+				return nil
+			}
+
+		lgr, _, err := provider.CreateFromSnapshot(snapshotDir)
+		require.NoError(t, err)
+		verifyIndexCreatedOnMarbleSize(lgr)
+	})
+
+	t.Run("create_indexes_on_couchdb_for_legacy_lifecycle", func(t *testing.T) {
+		snapshotDir, couchDBConfig, provider := setup()
+		defer func() {
+			require.NoError(t, statecouchdb.DropApplicationDBs(couchDBConfig))
+		}()
+
+		// mimic legacy chaincode "ns" installed and defiend and the package contains an index definition "sort index"
+		deployedCCInfoProvider := provider.initializer.DeployedChaincodeInfoProvider.(*mock.DeployedChaincodeInfoProvider)
+		deployedCCInfoProvider.AllChaincodesInfoReturns(
+			map[string]*ledger.DeployedChaincodeInfo{
+				"ns": {
+					Name:     "ns",
+					Version:  "version",
+					Hash:     []byte("hash"),
+					IsLegacy: true,
+				},
+				"anotherNs": {
+					Name:     "anotherNs",
+					Version:  "version",
+					Hash:     []byte("hash"),
+					IsLegacy: false,
+				},
+			},
+			nil,
+		)
+
+		installedChaincodeInfoProvider := &kvledgermock.ChaincodeInfoProvider{}
+		installedChaincodeInfoProvider.RetrieveChaincodeArtifactsReturns(
+			true, dbArtifactsBytes, nil,
+		)
+		cceventmgmt.Initialize(installedChaincodeInfoProvider)
+		lgr, _, err := provider.CreateFromSnapshot(snapshotDir)
+		require.NoError(t, err)
+
+		require.Equal(t, 1, installedChaincodeInfoProvider.RetrieveChaincodeArtifactsCallCount())
+		require.Equal(t,
+			&cceventmgmt.ChaincodeDefinition{
+				Name:    "ns",
+				Version: "version",
+				Hash:    []byte("hash"),
+			},
+			installedChaincodeInfoProvider.RetrieveChaincodeArtifactsArgsForCall(0),
+		)
+		verifyIndexCreatedOnMarbleSize(lgr)
+	})
+
+	t.Run("errors-propagation", func(t *testing.T) {
```

Comment:
Another nit, but why nest the other tests under this?

Reply:
My main purpose was to save the cost of invoking the `setup` function (which prepares a ledger, creates a snapshot as a starting point) each time for an error case tests. Otherwise, though I find that grouping of all error cases under an umbrella adds clarity but I am open to change this if you want me to over look the first part.

Reply:
Not a big deal